tools/config/config.go

```diff
@@ -1,70 +1,61 @@
 package config
 
 import (
-	"encoding/json"
 	"fmt"
-	clientauthenticationapi "k8s.io/client-go/pkg/apis/clientauthentication"
+	"net/http"
+	"net/url"
+
+	"k8s.io/client-go/pkg/apis/clientauthentication"
 	"k8s.io/client-go/plugin/pkg/client/auth/exec"
 	"k8s.io/client-go/rest"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
-	"net/http"
-	"net/url"
+	clientcmdv1 "k8s.io/client-go/tools/clientcmd/api/v1"
 	"sigs.k8s.io/cluster-inventory-api/apis/v1alpha1"
 )
 
-func getProviderFromClusterProfile(cluster *v1alpha1.ClusterProfile, providerName string) (v1alpha1.ProviderConfig, bool) {
-	for _, provider := range cluster.Status.CredentialProviders {
-		if provider.Name == providerName {
-			return provider.Config, true
-		}
-	}
-	return v1alpha1.ProviderConfig{}, false
+type Provider struct {
+	Name string
+	ExecConfig *clientcmdapi.ExecConfig
 }
 
-func configToExecCluster(config v1alpha1.ProviderConfig) (*clientauthenticationapi.Cluster, error) {
-	outConfig := &clientauthenticationapi.Cluster{
-		Server:                   config.Server,
-		TLSServerName:            config.TLSServerName,
-		InsecureSkipTLSVerify:    config.InsecureSkipTLSVerify,
-		CertificateAuthorityData: config.CertificateAuthorityData,
-		ProxyURL:                 config.ProxyURL,
-		DisableCompression:       config.DisableCompression,
-	}
-
-	return outConfig, nil
+type CredentialsProvider struct {
+	providers []Provider
 }
 
-// BuildConfigFromClusterProfile is to build the rest.Config to init the client.
-func BuildConfigFromClusterProfile(cluster *v1alpha1.ClusterProfile, providerName string) (*rest.Config, error) {
-	provider, found := getProviderFromClusterProfile(cluster, providerName)
-	if !found {
-		return nil, fmt.Errorf("no provider found for cluster profile %q", cluster.Name)
+func New(providers []Provider) *CredentialsProvider {
```

Comment:
could you give an example on how controller use this? How does the controller get providers and set here.

Reply:
a configuration file that we also define configuration API?

Reply:
the configuration file would live within the controller (likely mounted at runtime and would contain the []Provider).

Let me add a helper func